### PR TITLE
LVRUBYM-223: Users actions with Our day report. All back-end.

### PR DESCRIPTION
**Pull request by yaremakleputs**
## JIRA

* [Main JIRA ticket](http://ssu-jira.softserveinc.com/browse/LVRUBYM-223)
* [JIRA subtask](http://ssu-jira.softserveinc.com/browse/LVRUBYM-223)

## Code reviewers

- [x] @Mykhailo123 
- [x] @skrypnyk88 
- [x] @vikabelous 

### Second Level Review

- [ ] @omaik

## Summary of issue

Needs to create model, controller for our day report. Spec tests for it. Add routing. Add indexes.

## Summary of change

- Create model, controller for our day report. 
- Spec tests for it. 
- Add routing. 
- Add indexes.

## Testing approach

We do not need any tests here

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: app/controllers/v1/our_days_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+module V1
+  class OurDaysController < ApplicationController
+    before_action :find_groups
+    before_action :find_our_day, only: [:show, :update]
+
+    def index
+      @our_days = @group.our_days
```
Comment (skrypnyk88) [inline app/controllers/v1/our_days_controller.rb]: You will need add student scope for front end.

File: app/controllers/v1/our_days_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    def create
+      @our_day = @group.our_days.build(our_day_params)
+      if @our_day.save
+        render :show
+      else
+        head :bad_request
+      end
+    end
+
+    def show; end
```
Comment (skrypnyk88) [inline app/controllers/v1/our_days_controller.rb]: the same

File: app/views/v1/our_days/index.json.jbuilder [hunk 3]
```diff
@@ -0,0 +1,3 @@
+json.array! @our_days, :id,
```
Comment (skrypnyk88) [inline app/views/v1/our_days/index.json.jbuilder]: You must change it for client side

File: app/views/v1/our_days/show.json.jbuilder [hunk 4]
```diff
@@ -0,0 +1,3 @@
+json.(@our_day, :id,
```
Comment (skrypnyk88) [inline app/views/v1/our_days/show.json.jbuilder]: The same

File: db/migrate/20170410083300_add_group_reference_to_our_day.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddGroupReferenceToOurDay < ActiveRecord::Migration[5.0]
```
Comment (skrypnyk88) [inline db/migrate/20170410083300_add_group_reference_to_our_day.rb]: name of migration is incorrect

File: db/migrate/20170410155717_add_index_to_our_day.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToOurDay < ActiveRecord::Migration[5.0]
```
Comment (skrypnyk88) [inline db/migrate/20170410155717_add_index_to_our_day.rb]: the same

File: db/seeds.rb [hunk 7]
```diff
@@ -19,7 +19,11 @@
     birthdate: Faker::Date.birthday(2, 6),
     gender: Faker::Demographic.sex.downcase
   )
-
+  OurDay.create(
```
Comment (skrypnyk88) [inline db/seeds.rb]: Add your seeds too other people

File: app/controllers/v1/our_days_controller.rb (hunk 1 above)
Comment (yaremakleputs) [inline app/controllers/v1/our_days_controller.rb]: Our day report have not connection with students.

File: app/views/v1/our_days/index.json.jbuilder (hunk 3 above)
Comment (yaremakleputs) [inline app/views/v1/our_days/index.json.jbuilder]: Should not I write id?

File: app/controllers/v1/our_days_controller.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      @group = Group.find_by(id: params[:group_id])
+      head :bad_request unless @group
+    end
+
+    def find_our_day
+      @our_day = @group.our_days.find_by(id: params[:id])
+      head :bad_request unless @our_day
+    end
+
+    def render_show_or_bad_request(succeed)
```
Comment (VitaliaDurnyak) [inline app/controllers/v1/our_days_controller.rb]: Maybe,it's better to do like this - "succeed ? render :show  :  head :bad_request " ?)

File: app/controllers/v1/our_days_controller.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    def show; end
+
+    def update
+      render_show_or_bad_request(@our_day.update(our_day_params))
+    end
+
+    private
+
+    def find_groups
```
Comment (okohuch) [inline app/controllers/v1/our_days_controller.rb]: why do you have the plural form? you search just for one group

File: spec/controllers/v1/our_day_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+  let(:our_day) { create(:our_day, group: group) }
+
+  let(:report_attrs) { attributes_for(:our_day).with_indifferent_access }
+
+  let(:our_day_reports) do
+    3.times do
+      create(:our_day, group: group)
+    end
+    group
```
Comment (okohuch) [inline spec/controllers/v1/our_day_controller_spec.rb]: why do you return `group` here? I do not see any logic here

File: db/migrate/20170401170217_create_our_day.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class CreateOurDay < ActiveRecord::Migration[5.0]
```
Comment (omaik) [inline db/migrate/20170401170217_create_our_day.rb]: This migration is not named by convention

File: app/controllers/v1/our_days_controller.rb (hunk 9 above)
Comment (yaremakleputs) [inline app/controllers/v1/our_days_controller.rb]: I will change it. Thank you.

File: db/migrate/20170401170217_create_our_day.rb (hunk 11 above)
Comment (yaremakleputs) [inline db/migrate/20170401170217_create_our_day.rb]: Ok. I will change.

File: app/controllers/v1/our_days_controller.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+    def create
+      @our_day = @group.our_days.build(our_day_params)
+      render_show_or_bad_request(@our_day.save)
+    end
+
+    def show; end
+
+    def update
+      render_show_or_bad_request(@our_day.update(our_day_params))
```
Comment (okohuch) [inline app/controllers/v1/our_days_controller.rb]: It seems that we agreed to delete such methods. Didn't we?????

File: app/controllers/v1/our_days_controller.rb (hunk 8 above)
Comment (okohuch) [inline app/controllers/v1/our_days_controller.rb]: @yaremakleputs  Please remove this method and put the code in appropriate places.

File: config/locales/en.yml [hunk 13]
```diff
@@ -218,4 +218,4 @@ en:
       default: "%a, %d %b %Y %H:%M:%S %z"
       long: "%B %d, %Y %H:%M"
       short: "%d %b %H:%M"
-    pm: pm
+    pm: pm
```
Comment (Mykhailo123) [inline config/locales/en.yml]: Empty line

File: config/locales/ua.yml [hunk 14]
```diff
@@ -271,4 +271,4 @@ ua:
       default: "%a, %d %b %Y, %H:%M:%S %z"
       long: "%d %B %Y, %H:%M"
       short: "%d %b, %H:%M"
-    pm: по полудні
+    pm: по полудні
```
Comment (Mykhailo123) [inline config/locales/ua.yml]: Empty line

File: app/controllers/v1/our_days_controller.rb (hunk 12 above)
Comment (yaremakleputs) [inline app/controllers/v1/our_days_controller.rb]: Yes, I done it.

File: app/controllers/v1/our_days_controller.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  class OurDaysController < ApplicationController
+    include Reportable
+
+    def show
+      @our_day = single_report_renderer(OurDay, @group).call if @students.any?
+    end
+
+    def update
+      @our_day = OurDay.find_by(day: @day, group_id: @group)
+      if @our_day.update(our_day_params)
```
Comment (omaik) [inline app/controllers/v1/our_days_controller.rb]: update_attributes

File: app/views/v1/our_days/show.json.jbuilder [hunk 16]
```diff
@@ -0,0 +1,10 @@
+json.students(@group.students, :first_name,
```
Comment (omaik) [inline app/views/v1/our_days/show.json.jbuilder]: Use Vika's partial instead

File: config/routes.rb [hunk 17]
```diff
@@ -6,6 +6,7 @@
 
   namespace :v1 do
     resources :groups do
+      resources :students, shallow: true
```
Comment (omaik) [inline config/routes.rb]: any need for this?

File: spec/controllers/v1/our_day_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+  let(:students_attrs) do
+    students.map do |student|
+      student.attributes
+             .with_indifferent_access
+             .extract!(:first_name, :last_name)
+    end
+  end
+
+  let!(:our_day_present) do
```
Comment (omaik) [inline spec/controllers/v1/our_day_controller_spec.rb]: `let!` is redundant here and below

File: spec/controllers/v1/our_day_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    context 'when our_day is not valid' do
+      let(:invalid_params) do
+        { method: :patch,
+          day: Date.today,
+          group_id: group,
+          our_day: { description: 'a' * 1001 } }
+      end
+
+      it 'renders bad_reqest' do
+        post :update, format: :json,
```
Comment (omaik) [inline spec/controllers/v1/our_day_controller_spec.rb]: ouch

Comment (omaik) [inline spec/controllers/v1/our_day_controller_spec.rb]: Look at line 88

File: app/views/v1/our_days/show.json.jbuilder (hunk 16 above)
Comment (yaremakleputs) [inline app/views/v1/our_days/show.json.jbuilder]: When I pushed this commit, Vika's partial was not merged.

File: config/routes.rb [hunk 20]
```diff
@@ -24,5 +24,7 @@
     end
 
     resources :teachers
+    get '/our_day/:day', to: 'our_days#show'
```
Comment (omaik) [inline config/routes.rb]: use resources instead

File: spec/controllers/v1/our_day_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+        get :show, format: :json, params: { group_id: group, day: Date.today }
+        expect(response.body).to eq(our_day_absent.to_json)
+      end
+    end
+  end
+
+  describe 'PUT #update' do
+    context 'when our_day is valid' do
+      let(:valid_params) do
+        { method: :patch,
```
Comment (omaik) [inline spec/controllers/v1/our_day_controller_spec.rb]: you do not need this

File: spec/controllers/v1/our_day_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        put :update, format: :json,
+                     params: valid_params
+
+        expect(response.body).to eq(updated_our_day.to_json)
+      end
+    end
+
+    context 'when our_day is not valid' do
+      let(:invalid_params) do
+        { method: :patch,
```
Comment (omaik) [inline spec/controllers/v1/our_day_controller_spec.rb]: same here

File: config/routes.rb [hunk 23]
```diff
@@ -25,5 +25,6 @@
     resources :bottle_reports, only: [:index] do
       resources :bottles
     end
+    resource :our_day
```
Comment (omaik) [inline config/routes.rb]: do you need all list of routes?

**Pull request merged**